### PR TITLE
Support GICv2 in testing KVM_CREATE_DEVICE.

### PR DESCRIPTION
**Pull request by michael2012z**
On Aarch64, retry to create VGICv2 device after failing to create
VGICv3, in case GICv2 is equipped (like RaspBerry Pi 4).
Divide test_create_device() into 2 architecture specific functions,
to avoid too much confusing conditional compiling sentences.

Signed-off-by: Michael Zhao <michael.zhao@arm.com>

### CODE REVIEW

**Review by andreeaflorescu — DISMISSED** [review]

File: src/ioctls/device.rs
Comment (acatangiu) [inline src/ioctls/device.rs]: Why do this? `device_fd` is already a `DeviceFd`. Code should be equivalent without these 3 lines playing with the `unsafe` underlying `RawFd`.

File: src/ioctls/device.rs
Comment (acatangiu) [inline src/ioctls/device.rs]: These 2 lines can be collapsed into:
```rust
let device_fd = match vm.create_device(&mut gic_device) {
```

Comment (michael2012z) [reply src/ioctls/device.rs]: Thanks.

File: src/ioctls/device.rs
Comment (michael2012z) [reply src/ioctls/device.rs]: Yes, these 3 lines are unnecessary. I introduced them when splitting from the old test function. :)
In x86 test case, the `raw fd` was not needed either. I deleted them as well.

File: coverage_config.json
Comment (michael2012z) [inline coverage_config.json]: I removed 3 unneeded code lines in x86, so the coverage degraded a little bit.

File: src/ioctls/device.rs
Comment (acatangiu) [inline src/ioctls/device.rs:97]: The code-coverage decrease prompted me to check and find out that these 3 lines in this particular test were there to also test (as coverage drop suggests) `DeviceFd::as_raw_fd()` and `DeviceFd::from_raw_fd()`.

I suggest you either:

- bring back those lines in this test only and properly document them with comments explaining what their purpose is in the test,
or
- add another test that covers `DeviceFd::as_raw_fd()` and `DeviceFd::from_raw_fd()`.

**Review by acatangiu — CHANGES_REQUESTED** [review]
Decrease of coverage happens because `DeviceFd::as_raw_fd()` and `DeviceFd::from_raw_fd()` are no longer tested.

See inline comment.

File: src/ioctls/device.rs
Comment (michael2012z) [reply src/ioctls/device.rs:97]: Aha, yes, `as_raw_fd()` and `from_raw_fd()` are implemented for `DeviceFd`. Although they were not needed in our test cases, they should be tested.
I prefer to cover them in this test.

**Comment by michael2012z** [comment]
An error was triggered in ARM unit tests:

> ---- ioctls::vm::tests::test_irq_chip stdout ------
> thread 'ioctls::vm::tests::test_irq_chip' panicked at 'assertion failed: vm.create_irq_chip().is_err()', src/ioctls/vm.rs:1153:13

Creating a device by `vm.create_irq_chip()` on a machine with GICv3 is supposed to fail. That is what the case asserted.
But if the GIC is v2, this operation will succeed. I think that's why this case failed: the job was deployed on a GICv2 machine (maybe a RPI4?).

**Comment by michael2012z** [comment]
IMO, this line should be removed: https://github.com/rust-vmm/kvm-ioctls/blob/master/src/ioctls/vm.rs#L1153 .
And let `test_irq_chip()` be dedicated to X86 (skip this test for arm). Because:
1. For arm, this test is not a functioning one. It only asserted a failure case.
2. It's hardware dependent. On GIC v2 and v3, the result is totally different.

What do you think? @acatangiu

**Comment by andreeaflorescu** [comment]
Hey @michael2012z the error is not related to your PR.

I changed the bare metal instance on which we run the Arm CI and in the process, I also upgraded the kernel version. These changes are triggering the faiIures you are seeing. I will fix the failures in a PR, but we should wait for the fixes to be merged before we can merge your PR. I'll keep you posted.

**Comment by michael2012z** [comment]
Thank you, @andreeaflorescu .

**Review by acatangiu — APPROVED** [review]

**Comment by michael2012z** [comment]
@andreeaflorescu Thank you for clearing the error with https://github.com/rust-vmm/kvm-ioctls/pull/78. 
Now all checks passed.

**Review by andreeaflorescu — APPROVED** [review]

**Pull request merged**
